File: app/src/main/java/de/devmil/paperlaunch/config/LaunchConfig.java
/*
 * Copyright 2015 Devmil Solutions
 *
 * Licensed under the Apache License, Version 2.0 (the "License");
 * you may not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 *
 *     http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 */
package de.devmil.paperlaunch.config;

import android.graphics.Color;

import java.util.ArrayList;
import java.util.List;

import de.devmil.paperlaunch.model.IEntry;

public class LaunchConfig implements ILaunchEntryConfig, ILaunchLaneConfig {
    private List<IEntry> mEntries;
    private float mImageWidthDip = 48;
    private float mImageMarginDip = 1;
    private float mEntriesMarginDip = 3;
    private float mLowerElevationDip = 2;
    private float mHighElevationDip = 6;
    private int mEntryMoveAnimationDuration = 100;
    private int mEntryMoveDiffMS = 50;
    private int mEntryAlphaAnimationDuration = 100;
    private int mSelectingAnimationDurationMS = 200;
    private float mItemNameTextSizeSP = 30;
    private IDesignConfig mDesignConfig = new DesignConfig();
    private float mNeutralZoneWidthDip = 50;
    private int mLauncherInitAnimationDurationMS = 150;
    private float mImageOffsetDip = 5;
    private float mLaneIconTopMarginDip = 5;
    private float mLaneTextTopMarginDip = 20;
    private int mLauncherBackgroundColor = Color.BLACK;
    private float mLauncherBackgroundAlpha = 0.3f;
    private int mLauncherBackgroundAnimationDurationMS = 250;

    //UserSettings
    private int mLauncherSensitivityDip;
    private boolean mIsOnRightSide;
    private int mLauncherOffsetPositionDip;
    private int mLauncherOffsetHeightDip;
    private boolean mShowLauncherBackground;
    private boolean mIsVibrateOnActivation;

    public List<IEntry> getEntries()
    {
        return new ArrayList<>(mEntries);
    }

    public LaunchConfig(IUserSettings userSettings)
    {
        mLauncherSensitivityDip = userSettings.getSensitivityDip();
        mIsOnRightSide = userSettings.isOnRightSide();
        mLauncherOffsetPositionDip = userSettings.getActivationOffsetPositionDip();
        mLauncherOffsetHeightDip = userSettings.getActivationOffsetHeightDip();
        mShowLauncherBackground = userSettings.isShowBackground();
        mIsVibrateOnActivation = userSettings.isVibrateOnActivation();
    }

    public void setEntries(List<IEntry> entries)
    {
        mEntries = new ArrayList<>(entries);
    }

    public float getImageWidthDip()
    {
        return mImageWidthDip;
    }

    public float getImageMarginDip() {
        return mImageMarginDip;
    }

    public float getEntriesMarginDip() {
        return mEntriesMarginDip;
    }

    public float getLowElevationDip()
    {
        return mLowerElevationDip;
    }

    public float getHighElevationDip()
    {
        return mHighElevationDip;
    }

    public boolean isOnRightSide()
    {
        return mIsOnRightSide;
    }

    public int getSelectingAnimationDurationMS() {
        return mSelectingAnimationDurationMS;
    }

    public float getItemNameTextSizeSP() {
        return mItemNameTextSizeSP;
    }

    public int getEntryMoveAnimationDuration()
    {
        return mEntryMoveAnimationDuration;
    }

    public int getEntryAlphaAnimationDuration()
    {
        return mEntryAlphaAnimationDuration;
    }

    public int getEntryMoveDiffMS()
    {
        return mEntryMoveDiffMS;
    }

    public IDesignConfig getDesignConfig()
    {
        return mDesignConfig;
    }

    public float getNeutralZoneWidthDip() {
        return mNeutralZoneWidthDip;
    }

    public int getLauncherInitAnimationDurationMS() {
        return mLauncherInitAnimationDurationMS;
    }

    public float getImageOffsetDip()
    {
        return mImageOffsetDip;
    }

    public float getLaneIconTopMarginDip() {
        return mLaneIconTopMarginDip;
    }

    public float getLaneTextTopMarginDip() {
        return mLaneTextTopMarginDip;
    }

    public float getLauncherBackgroundAlpha() {
        return mLauncherBackgroundAlpha;
    }
    public int getLauncherBackgroundAnimationDurationMS() {
        return mLauncherBackgroundAnimationDurationMS;
    }

    public boolean getShowLauncherBackground() {
        return mShowLauncherBackground;
    }

    public int getLauncherBackgroundColor() {
        return mLauncherBackgroundColor;
    }

    public int getLauncherSensitivityDip() {
        return mLauncherSensitivityDip;
    }

    public int getLauncherOffsetPositionDip() {
        return mLauncherOffsetPositionDip;
    }

    public int getLauncherOffsetHeightDip() {
        return mLauncherOffsetHeightDip;
    }

    public boolean isVibrateOnActivation() {
        return mIsVibrateOnActivation;
    }
}

File: app/src/main/java/de/devmil/paperlaunch/config/LaunchConfig.kt
/*
 * Copyright 2015 Devmil Solutions
 *
 * Licensed under the Apache License, Version 2.0 (the "License");
 * you may not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 *
 *     http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 */
package de.devmil.paperlaunch.config

import android.graphics.Color

import java.util.ArrayList

import de.devmil.paperlaunch.model.IEntry

class LaunchConfig(userSettings: IUserSettings) : ILaunchEntryConfig, ILaunchLaneConfig {
    private var mEntries: List<IEntry>? = null

    private val BASE_IMAGE_WIDTH_DIP = 48f

    override val imageMarginDip = 1f
    override val entriesMarginDip = 3f
    override val lowElevationDip = 2f
    override val highElevationDip = 6f
    override val entryMoveAnimationDuration = 100
    override val entryMoveDiffMS = 50
    override val entryAlphaAnimationDuration = 100
    override val selectingAnimationDurationMS = 200
    override val itemNameTextSizeSP = 30f
    override val designConfig: IDesignConfig = DesignConfig()
    val neutralZoneWidthDip = 50f
    val launcherInitAnimationDurationMS = 150
    override val imageOffsetDip = 5f
    override val laneIconTopMarginDip = 5f
    override val laneIconMarginsDip = 3f
    override val laneTextTopMarginDip = 20f
    val launcherBackgroundColor = Color.BLACK
    val launcherBackgroundAlpha = 0.3f
    val launcherBackgroundAnimationDurationMS = 250
    val maxFolderDepth = 9
    val launcherHintAlpha = 0.28f

    //UserSettings
    @Suppress("JoinDeclarationAndAssignment")
    val launcherSensitivityDip: Int
    override val isOnRightSide: Boolean
    val showLauncherBackground: Boolean
    val isVibrateOnActivation: Boolean
    val launcherOffsetPosition: Int
    val launcherHeightPercent: Int
    val showLauncherHint : Boolean
    override val launcherGravity: LauncherGravity?
    override val imageWidthDip: Float
    override val showLogo: Boolean

    var entries: List<IEntry>
        get() = mEntries?.let { ArrayList(it) } ?: ArrayList()
        set(entries) {
            mEntries = ArrayList(entries)
        }

    init {
        launcherSensitivityDip = userSettings.sensitivityDip
        isOnRightSide = userSettings.isOnRightSide
        showLauncherBackground = userSettings.isShowBackground
        isVibrateOnActivation = userSettings.isVibrateOnActivation
        launcherGravity = userSettings.launcherGravity
        imageWidthDip = (BASE_IMAGE_WIDTH_DIP * userSettings.itemScalePercent) / 100f
        showLogo = userSettings.showLogo
        showLauncherHint = userSettings.isShowHint
        launcherOffsetPosition = userSettings.activationOffsetPosition
        launcherHeightPercent = userSettings.activationHeightPercent
    }
}
